### PR TITLE
Networking and plumbing enhancements

### PR DESCRIPTION
**Pull request by martinsawicki**
- completing support for NSG security rule
- Network#subnets() as Map
- pushing service manager tracking into base impl classes
- Subnet.networkSecurityGroup() support

### CODE REVIEW

File: azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/NetworkSecurityRule.java [hunk 1]
```diff
@@ -177,182 +182,196 @@ public static Protocol fromString(String s) {
...
+        }
 
-    /**
-     * Grouping of security rule update stages.
-     */
-    interface Updatables {
         /**
-         * The stage of the network rule description allowing the direction and the access type to be specified.
+         * The stage of the security rule definition allowing the protocol that the rule applies to to be specified.
+         * @param <ParentT> the return type of the final {@link DefinitionAttachable#attach()}
```
Comment (anuchandy) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/NetworkSecurityRule.java]: `{@ Attachable#attach()}` ?

Comment (martinsawicki) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/NetworkSecurityRule.java]: yeah, I wish we could make checkstyle check for bad references in Javadocs...

Comment (anuchandy) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/NetworkSecurityRule.java]: At least in IntelliJ bad references are getting highlighted by editor

Comment (anuchandy) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/NetworkSecurityRule.java]: We can enable this in eclipse http://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Freference%2Fpreferences%2Fjava%2Fcompiler%2Fref-preferences-javadoc.htm


Preferences -> Java -> Compiler -> Javadoc

Check "Process Javadoc comments"
Set "Malformed Javadoc comments" as "Warning"
Check "Validate tag arguments"

Comment (martinsawicki) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/NetworkSecurityRule.java]: works

**Pull request merged**